### DISCUSSION TITLE
Access request headers from `events` callbacks

### DISCUSSION THREAD

**Discussion by michaelhays** [nextauthjs/next-auth#8991] [Help] [answered]
We use the [`signIn()` callback](https://authjs.dev/guides/basics/events#signin) to record audit logs, and need to access the request headers to record relevant fields like the user's IP address.

Is this possible in NextAuth.js? I'm using version `5.0.0-beta.3`.

**Comment by Wundero** [comment, marked as answer]
If you are using the app router, I believe you can get headers using the `headers()` function from `next/headers`. Besides that, there doesn't appear to be a way to access the headers from that callback. 

Below is the reference in the code:
![image](https://github.com/nextauthjs/next-auth/assets/6076081/59cd8106-5ec2-4b8e-a345-a17d8e46c905)
[OAuth signIn callback found here](https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/lib/routes/callback.ts#L168).
If you are willing to delve into your `node_modules` and use something like [`patch-package`](https://www.npmjs.com/package/patch-package), you could make the changes necessary yourself (the function that calls the callbacks does have access to headers), but this might be a bit tricky.

I think it could be a good addition to the library, since having access to headers (and cookies, for that matter) in the callbacks might make certain tasks easier.

**Reply by michaelhays** [reply]
Wow, `headers()` worked perfectly :tada:

I didn't even bother trying that before because I thought it would've been outside of that context, hah. Thanks so much!

**Reply by YHADJRABIA** [reply]
Found a simpler way with  `cookies()` of `next/headers`
      
```
      const cookies = cookies()
      const locale = cookies?.get('NEXT_LOCALE')?.value
```

It gives me access to the locale with the `signIn` callback

**Reply by heijmerikx** [reply]
Yeah, great! Just what I was looking for as well. I didn't thought about this solution too. I tried wrapping the NextAuth config in a request, which did seem to work, but got a `TypeError: r is not a function` on deployment. This works like charm! A year later on Next.js v14 using auth.js beta.
